Assets/Scripts/BoardManager.cs

```diff
@@ -22,47 +44,168 @@ public class BoardManager : MonoBehaviour
     private Player player;
     private Transform boardHolder;
     private List<Vector3> gridPositions = new List<Vector3>();
+    private List<Vector3> outerWallGridPositions = new List<Vector3>();
+    private List<Vector3> innerWallGridPositions = new List<Vector3>();
 
-    void InitialiseList()
+    private int numberOfRoom = 1;
+
+    void BoardSetup()
     {
-        gridPositions.Clear();
+        columnsCount = Random.Range(columns.minimum, columns.maximum + 1);
+        rowsCount = Random.Range(rows.minimum, rows.maximum + 1);
+
+        numberOfRoom = Random.Range(numberOfRoomParts.minimum, numberOfRoomParts.maximum + 1) - 1;
+
+        boardHolder = new GameObject("Board").transform;
+
+        int x = 0;
+        int y = 0;
+        List<string> roomPosition = new List<string>();
+        List<Vector2> m = new List<Vector2>() { new Vector2(x, y) };
+        List<string> validPosition = new List<string>() { "top", "right", "bottom" };
+
+        //Calculate layout of other parts of the room
+        for (int i = 0; i < numberOfRoom; i++)
+        {
+            int randomIndex = Random.Range(0, 3);
+            if (validPosition[randomIndex] == "top")
+            {
+                roomPosition.Add("top");
+                y++;
+                m.Add(new Vector2(x, y));
+                validPosition = new List<string>() { "top", "right", "left" };
+            }
+            else if (validPosition[randomIndex] == "right")
+            {
+                roomPosition.Add("right");
+                x++;
+                m.Add(new Vector2(x, y));
+                validPosition = new List<string>() { "top", "right", "bottom" };
+            }
+            else if (validPosition[randomIndex] == "bottom")
+            {
+                roomPosition.Add("bottom");
+                y--;
+                m.Add(new Vector2(x, y));
+                validPosition = new List<string>() { "right", "bottom", "left" };
+            }
+            else if (validPosition[randomIndex] == "left")
+            {
+                roomPosition.Add("left");
+                x--;
+                m.Add(new Vector2(x, y));
+                validPosition = new List<string>() { "top", "bottom", "left" };
+            }
+        }
+
+        //Generate first room part
+        GenerateRoomPart(x, y, m, 0);
 
-        for (int x = 1; x < columns - 1; x++)
+        //Generate other room parts
+        int index = 1;
+        foreach (var pos in roomPosition)
         {
-            for (int y = 1; y < rows - 1; y++)
+            if (pos == "top")
```

Comment:
I'm not sure if Switch cases are possible with strings in C# but it is with enum types ;)